### ISSUE TITLE
Segmentation fault with parsing of alignment at 'Compressing aligmnent patterns...'

### ISSUE THREAD

**Issue by jgolob** [amkozlov/raxml-ng#122]
Greetings! Thank you for the well-supported re-implementation of raxml. 

When attempting to parse an alignment, I am consistently running into a segmentation fault with V1.0.2. 
This has reproduced with the same alignment on multiple systems / setups. `--check` is able to complete successfully.

The log (with `--log debug`):
```
RAxML-NG v. 1.0.2 released on 22.02.2021 by The Exelixis Lab.
Developed by: Alexey M. Kozlov and Alexandros Stamatakis.
Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth.
Latest version: https://github.com/amkozlov/raxml-ng
Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml

System: Intel(R) Core(TM) i7-8086K CPU @ 4.00GHz, 6 cores, 31 GB RAM

RAxML-NG was called at 22-Jul-2021 18:15:07 as follows:

raxml-ng --parse --msa ra.fasta --seed 12345 --model GTR+G --log debug

Analysis options:
  run mode: Alignment parsing and compression
  start tree(s): 
  random seed: 12345
  tip-inner: OFF
  pattern compression: ON
  per-rate scalers: OFF
  site repeats: ON
  branch lengths: proportional (ML estimate, algorithm: NR-FAST)
  SIMD kernels: AVX2
  parallelization: coarse-grained (auto), PTHREADS (auto)

RBA partial loading: OFF
|noname|   |GTR+FO+G4m|   ||
[00:00:00] Reading alignment from file: ra.fasta
Failed to load as IPHYLIP: Unable to parse PHYLIP file: ra.fasta
 (LIBPLL-231): Invalid number of sequences in header
Failed to load as PHYLIP: Unable to parse PHYLIP file: ra.fasta
 (LIBPLL-231): Invalid number of sequences in header
[00:00:01] Loaded alignment with 46991 taxa and 12777 sites
[00:00:01] Extracting partitions... 
[00:00:01] Checking the alignment...

WARNING: Fully undetermined columns found: 60

NOTE: Reduced alignment (with duplicates and gap-only sites/taxa removed) 
NOTE: was saved to: /working/ra.fasta.raxml.reduced.phy
[00:00:47] Compressing alignment patterns... 
Segmentation fault
```

The alignment is gigantic (~ 500mb) making it a bit impractical to share directly.

**Comment by amkozlov** [comment]
Hello, thank you for reporting! But i'm afraid it will be difficult to diagnose this problem without having the MSA file. Could you please share it via GDrive etc.?

On a different note: most likely, it will make little sense to run raxml-ng on a dataset with such dimensions (#taxa > #sites) due to insufficient signal. We have discussed this issue quite a lot on the raxml google group, see e.g. these threads:
https://groups.google.com/g/raxml/c/13MxNQvve-Y/m/4rLAdj3MAAAJ
https://groups.google.com/g/raxml/c/upsTachi-Nc/m/etOL1bW0BAAJ

**Comment by amkozlov** [comment]
If the issue still exists with the latest raxml-ng version, please reopen and provide input files.

**Comment by domivika** [comment]
Hello,

I managed to solve this issue. I had to change the stack size and open files limitations on the server. 
These settings worked for me, in case somebody encounters the same problem:

ulimit -n 65536
ulimit -s unlimited

**Comment by amkozlov** [comment]
Hi Dominika,

thanks for sharing the solution!
I was just about to reply since the same issue was reported a couple of years ago on the google group:
https://groups.google.com/g/raxml/c/fdyUVMkONkk/m/1FlImtJbAQAJ

Please note however that ML tree inference with 50k+ taxa could take _a lot_ of time. so consider suggestions in the google group thread above.

**Comment by domivika** [comment]
Hi Oleksiy,

Thanks, I wasn't aware of the google group, my bad.

I have one more question regarding the performance. I have 56 cores in total (28 physical and 56 logical) and 377 GB of RAM. When running raxml-ng with --parse, I got this information:
` 
* Estimated memory requirements                : 19164 MB
* Recommended number of threads / MPI processes: 5
`

So, I am running the job with such parameters right now:
raxml-ng --msa results/fasta/raxml-ready.fa --model GTR+G --tree-constraint results/fasta/raxml-ready.tre --log DEBUG --search --threads auto{56} --workers 10

Log file:
`
RAxML-NG v. 1.2.2 released on 11.04.2024 by The Exelixis Lab.
Developed by: Alexey M. Kozlov and Alexandros Stamatakis.
Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth, Julia Haag, Anastasis Togkousidis.
Latest version: https://github.com/amkozlov/raxml-ng
Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml

System: Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz, 28 cores, 377 GB RAM

RAxML-NG was called at 04-Mar-2025 17:11:51 as follows:

raxml-ng --msa results/fasta/raxml-ready.fa --model GTR+G --tree-constraint results/fasta/raxml-ready.tre --log DEBUG --search --threads auto{56} --workers 10

Analysis options:
  run mode: ML tree search
  start tree(s): random (10) + parsimony (10)
  topological constraint: results/fasta/raxml-ready.tre (algorithm: NEW)
  random seed: 1741104711
  tip-inner: OFF
  pattern compression: ON
  per-rate scalers: OFF
  site repeats: ON
  logLH epsilon: general: 10.000000, brlen-triplet: 1000.000000
  fast spr radius: AUTO
  spr subtree cutoff: 1.000000
  fast CLV updates: ON
  branch lengths: proportional (ML estimate, algorithm: NR-FAST)
  SIMD kernels: AVX2
  parallelization: coarse-grained (10 workers), PTHREADS (auto)

RBA partial loading: OFF
WARNING: The model you specified on the command line (GTR+G) will be ignored 
         since the binary MSA file already contains a model definition.
         If you want to change the model, please re-run RAxML-NG 
         with the original PHYLIP/FASTA alignment and --redo option.

[00:00:00] Loading binary alignment from file: results/fasta/raxml-ready.fa.raxml.rba
[00:00:00] Alignment comprises 57887 taxa, 1 partitions and 1356 patterns

Partition 0: noname
Model: GTR+FO+G4m
Alignment sites / patterns: 1690 / 1356
Gaps: 62.37 %
Invariant sites: 41.01 %


|noname|   |GTR+FO+G4m|   |1-1718|
[00:00:00] Loading constraint tree from: results/fasta/raxml-ready.tre
[00:00:00] Loaded non-comprehensive constraint tree with 8009 taxa

Recommended threads (response/balanced/throughput): 9 / 5 / 1

Parallelization scheme autoconfig: 10 worker(s) x 5 thread(s)


NOTE: Per-rate scalers were automatically enabled to prevent numerical issues on taxa-rich alignments.
NOTE: You can use --force switch to skip this check and fall back to per-site scalers.

[00:00:00] Generating a RANDOM starting tree, seed: 225333442
Parallel reduction/worker buffer size: 1 KB  / 0 KB

NOTE: Loaded checkpoint file: results/fasta/raxml-ready.fa.raxml.ckp
[00:00:04] NOTE: Resuming execution from checkpoint (logLH: -17831844.29, ML trees: 0, bootstraps: 0)

Initial model parameters:
   Partition: noname
   Rate heterogeneity: GAMMA (4 cats, mean),  alpha: 1.000000 (ML),  weights&rates: (0.250000,0.136954) (0.250000,0.476752) (0.250000,1.000000) (0.250000,2.386294) 
   Base frequencies (ML): 0.250000 0.250000 0.250000 0.250000 
   Substitution rates (ML): 1.000000 1.000000 1.000000 1.000000 1.000000 1.000000 

[00:00:04] Data distribution: max. partitions/sites/weight per thread: 1 / 272 / 4352

thread#	part#	start	length	weight
0	0	1088	268	4288

1	0	816	272	4352

2	0	544	272	4352

3	0	272	272	4352

4	0	0	272	4352

[00:00:04] Data distribution: max. searches per worker: 2

Starting ML tree search with 20 distinct starting trees

[00:01:01 -17831844.286499] AUTODETECT spr round 1 (radius: 5)
`

I just want to understand better what is going on right now. Since there are 20 starting trees and I have 10 workers with 2 max searches per worker, does it mean that the current process is being executed on all the 20 trees simultaneously? 

Thanks!

**Comment by amkozlov** [comment]
Hi Dominika,

yes, in your latest run with `--workers 10`, there are 10 tree search from 10 distinct starting trees running in parallel.

This parallelization setup should work fine. However, logical CPU cores are often not beneficial for raxml-ng, so you might as well try `--threads auto{28}`. Which option is better/faster is hard to tell without trying.
